File: y2024/day15/day15.py
from dataclasses import dataclass
from enum import Enum


class Direction(Enum):
    up = "^"
    down = "v"
    left = "<"
    right = ">"


@dataclass
class Coordinate:
    row: int
    col: int

    def move(self, direction: Direction) -> "Coordinate":
        ...


class Warehouse:
    def __init__(self, layout: list[list[str]]):
        self.layout: list[list[str]] = layout
        self.robot_pos: Coordinate = None

    def get_robot_pos(self) -> Coordinate:
        if self.robot_pos is not None:
            return self.robot_pos
        for row_num, row in self.layout:
            try:
                robot_col = row.index("@")
                self.robot_pos = Coordinate(row=row_num, col=robot_col)
                return self.robot_pos
            except IndexError:
                continue
        raise ValueError("robot not found!")

    def move_robot(self, direction: str) -> None:
        ...


def process_input() -> tuple[Warehouse, list[Direction]]:
    # warehouse
    with open("./day15/input_warehouse.txt", "r") as fp:
        warehouse = fp.readlines()

    warehouse = [list(i.strip()) for i in warehouse]
    warehouse = Warehouse(layout=warehouse)

    # movements
    with open("./day15/input_robot_movements.txt", "r") as fp:
        movements = fp.readlines()
    movements = [i.strip() for i in movements]
    directions = []
    for i in "".join(movements):
        if i == "^":
            directions.append(Direction.up)
        elif i == "v":
            directions.append(Direction.down)
        elif i == "<":
            directions.append(Direction.left)
        else:
            directions.append(Direction.right)

    return warehouse, directions


def part01(warehouse: Warehouse, movements: str):
    ...


def part02(warehouse: Warehouse, movements: str):
    ...


def main():
    warehouse, movements = process_input()
    part01(warehouse, movements)
    part02(warehouse, movements)


if __name__ == "__main__":
    main()

File: y2024/day15/day15.go
package day15

import (
	"adventofcode/y2024/utils"
	"fmt"
	"log"
	"time"
)

const (
	ROBOT_INDICATOR = '@'
	CRATE           = 'O'
	EMPTY_SPACE     = '.'
	WALL            = '#'
)

type RobotPos struct {
	Row int
	Col int
}

func (r *RobotPos) Equal(other RobotPos) bool {
	return r.Row == other.Row && r.Col == other.Col
}

func moveRobot(direction rune, pos RobotPos) RobotPos {
	switch direction {
	case '^':
		return RobotPos{Row: pos.Row - 1, Col: pos.Col}
	case 'v':
		return RobotPos{Row: pos.Row + 1, Col: pos.Col}
	case '<':
		return RobotPos{Row: pos.Row, Col: pos.Col - 1}
	default:
		return RobotPos{Row: pos.Row, Col: pos.Col + 1}
	}
}

func showGrid(grid [][]rune) {
	for _, row := range grid {
		for _, val := range row {
			fmt.Printf("%c", val)
		}
		fmt.Println()
	}
}

func getRobotPosition(grid [][]rune) (RobotPos, error) {
	for rowNum, row := range grid {
		for colNum, val := range row {
			if val == ROBOT_INDICATOR {
				return RobotPos{Row: rowNum, Col: colNum}, nil
			}
		}
	}
	return utils.GetZero[RobotPos](), fmt.Errorf("could not find robot position")
}

func processInput(daydir string) ([][]rune, string) {
	moves := utils.Read(daydir + "/input_robot_movement.txt")

	warehouseLines := utils.ReadLines(daydir + "/input_warehouse.txt")
	grid := [][]rune{}
	for _, line := range warehouseLines {
		row := []rune{}
		for _, ch := range line {
			row = append(row, ch)
		}
		grid = append(grid, row)
	}

	return grid, moves
}

func adjustCrates(grid *[][]rune, pos RobotPos, direction rune) RobotPos {
	nextPos := moveRobot(direction, pos)
	nextCh := (*grid)[nextPos.Row][nextPos.Col]
	switch nextCh {
	case CRATE:
		if !nextPos.Equal(pos) {
			(*grid)[nextPos.Row][nextPos.Col] = CRATE
			(*grid)[pos.Row][pos.Col] = EMPTY_SPACE
		}
		return nextPos
	case WALL:
		return pos
	case EMPTY_SPACE:
		if (*grid)[pos.Row][pos.Col] == CRATE {
			(*grid)[nextPos.Row][nextPos.Col] = CRATE
			(*grid)[pos.Row][pos.Col] = EMPTY_SPACE
		}
		return nextPos
	}
	log.Fatalf("invalid 7character: %c", nextCh)
	return utils.GetZero[RobotPos]()
}

func part01(grid [][]rune, moves string) {
	robotPosition, err := getRobotPosition(grid)
	if err != nil {
		log.Fatalf("Could not find robot in grid!!")
	}
	for _, ch := range moves {
		nextRobotPos := moveRobot(ch, robotPosition)
		switch grid[nextRobotPos.Row][nextRobotPos.Col] {
		case '.':
			grid[robotPosition.Col][robotPosition.Col] = EMPTY_SPACE
			grid[nextRobotPos.Row][nextRobotPos.Col] = ROBOT_INDICATOR
			robotPosition = nextRobotPos
		case '#':
			continue
		case 'O':
			nextRobotPos = adjustCrates(&grid, robotPosition, ch)
			if !nextRobotPos.Equal(robotPosition) {
				grid[nextRobotPos.Row][nextRobotPos.Col] = ROBOT_INDICATOR
				grid[robotPosition.Row][robotPosition.Col] = EMPTY_SPACE
				robotPosition = nextRobotPos
			}
		}
		utils.ClearScreen()
		showGrid(grid)
		time.Sleep(500 * time.Millisecond)
	}
}

func Run(dir string) {
	grid, moves := processInput(dir + "/day15")
	part01(grid, moves)
}
